File: vibeshield-python/src/deserialization_protector.py
import json
import logging
import re
from typing import Any, Dict, List, Optional

logger = logging.getLogger("VibeShield")

class VibeShieldDeserializationError(ValueError):
    """Exception raised when a deserialization security issue is detected."""
    def __init__(self, message: str):
        super().__init__(f"[VibeShield] Deserialization Protection: {message}")
        self.message = message

def _check_json_depth(text: str, max_depth: int) -> None:
    """
    Iteratively checks the nesting depth of a JSON string without parsing it completely.

    Args:
        text: The raw JSON string.
        max_depth: The maximum nesting depth allowed.

    Raises:
        VibeShieldDeserializationError: If the maximum depth is exceeded.
    """
    depth = 0
    max_reached = 0
    in_string = False
    escape = False
    i = 0
    n = len(text)
    while i < n:
        c = text[i]
        if in_string:
            if escape:
                escape = False
            elif c == '\\':
                escape = True
            elif c == '"':
                in_string = False
        else:
            if c == '"':
                in_string = True
            elif c in ('{', '['):
                depth += 1
                if depth > max_reached:
                    max_reached = depth
                if max_reached > max_depth:
                    raise VibeShieldDeserializationError(f"Maximum JSON depth of {max_depth} exceeded.")
            elif c in ('}', ']'):
                depth -= 1
        i += 1

def safe_json_parse(text: str, max_depth: int = 64) -> Any:
    """
    Safely parses JSON string with depth validation and key sanitisation.

    Args:
        text: The JSON string to parse.
        max_depth: The maximum nesting depth allowed.

    Returns:
        The deserialized and sanitized JSON payload.

    Raises:
        VibeShieldDeserializationError: If depth limit is exceeded or parsing fails.
    """
    _check_json_depth(text, max_depth)
    try:
        parsed = json.loads(text)
        
        # Parity guard: clean up key words
        def clean_object(obj: Any) -> Any:
            if obj is None or not isinstance(obj, (dict, list)):
                return obj
            if isinstance(obj, list):
                return [clean_object(item) for item in obj]
            
            clean = {}
            for k, v in obj.items():
                if str(k) in ('__proto__', 'constructor', 'prototype'):
                    continue
                clean[k] = clean_object(v)
            return clean

        return clean_object(parsed)
    except Exception as e:
        if isinstance(e, VibeShieldDeserializationError):
            raise e
        raise VibeShieldDeserializationError(f"Malformed JSON payload: {str(e)}")

def detect_unsafe_deserialization(code: str) -> Dict[str, Any]:
    """
    Statically analyzes source code to detect dangerous deserialization patterns.

    Args:
        code: The Python source code string to analyze.

    Returns:
        A dict containing:
            - safe: A boolean indicating if no unsafe patterns were found.
            - findings: A list of dicts detailing each unsafe finding.
    """
    findings = []
    if not code:
        return {"safe": True, "findings": findings}

    lines = code.split('\n')
    def get_line_num(index: int) -> int:
        return len(code[:index].split('\n'))

    # Rules mapping regex to description
    unsafe_libs = [
        (r'\bpickle\.(loads|load)\b', 'pickle', 'critical', 'Avoid using pickle.load/loads as it allows remote code execution.'),
        (r'\bmarshal\.(loads|load)\b', 'marshal', 'critical', 'Avoid using marshal.load/loads for untrusted payloads.'),
        (r'\bshelve\.open\b', 'shelve', 'critical', 'Avoid using shelve for untrusted persistent storage.'),
        (r'\bdill\.(loads|load)\b', 'dill', 'critical', 'dill deserialization allows remote code execution.'),
        (r'\bjsonpickle\.decode\b', 'jsonpickle', 'critical', 'jsonpickle.decode is vulnerable to arbitrary code execution.'),
    ]

    for regex, name, severity, sug in unsafe_libs:
        pattern = re.compile(regex)
        for match in pattern.finditer(code):
            line_num = get_line_num(match.start())
            findings.append({
                "line": line_num,
                "pattern": name,
                "severity": severity,
                "suggestion": sug,
                "original_code": lines[line_num - 1].strip()
            })

    # Rule: yaml.load without SafeLoader
    yaml_pattern = re.compile(r'\byaml\.load\s*\(')
    for match in yaml_pattern.finditer(code):
        line_num = get_line_num(match.start())
        # check if SafeLoader is passed
        # Safe loaders in PyYAML: SafeLoader or safe_load
        # If Loader=yaml.Loader or Loader=Loader is passed, it is unsafe.
        # We check the rest of the line or surrounding context
        context = lines[line_num - 1]
        is_safe = 'SafeLoader' in context or 'safe_load' in context
        if not is_safe:
            findings.append({
                "line": line_num,
                "pattern": "yaml.load without SafeLoader",
                "severity": "critical",
                "suggestion": "Use yaml.safe_load() or specify Loader=yaml.SafeLoader to prevent RCE.",
                "original_code": context.strip()
            })

    findings.sort(key=lambda x: x['line'])
    return {"safe": len(findings) == 0, "findings": findings}

File: vibeshield-ts/src/core/deserialization-protector.ts
/**
 * Error class thrown when a deserialization or prototype pollution violation occurs.
 */
export class VibeShieldDeserializationError extends Error {
  /**
   * Creates an instance of VibeShieldDeserializationError.
   * @param message - The error message.
   */
  constructor(message: string) {
    super(`[VibeShield] Deserialization Protection: ${message}`);
    this.name = 'VibeShieldDeserializationError';
    Object.setPrototypeOf(this, VibeShieldDeserializationError.prototype);
  }
}

/**
 * Finding representing a security vulnerability detected by static analysis.
 */
export interface Finding {
  line: number;
  pattern: string;
  severity: 'critical' | 'high' | 'warning';
  suggestion: string;
  originalCode: string;
}

/**
 * Detection result containing safety assessment and specific findings.
 */
export interface DetectionResult {
  safe: boolean;
  findings: Finding[];
}

/**
 * Checks the depth of JSON structures in a text payload.
 * @param text - The raw JSON string.
 * @param maxDepth - The maximum allowed nesting depth.
 * @throws VibeShieldDeserializationError if depth limit is exceeded.
 */
function checkJsonDepth(text: string, maxDepth: number): void {
  let depth = 0;
  let maxReached = 0;
  let inString = false;
  let escape = false;

  for (let i = 0; i < text.length; i++) {
    const c = text[i];
    if (inString) {
      if (escape) {
        escape = false;
      } else if (c === '\\') {
        escape = true;
      } else if (c === '"') {
        inString = false;
      }
    } else {
      if (c === '"') {
        inString = true;
      } else if (c === '{' || c === '[') {
        depth++;
        if (depth > maxReached) maxReached = depth;
        if (maxReached > maxDepth) {
          throw new VibeShieldDeserializationError(`Maximum JSON depth of ${maxDepth} exceeded.`);
        }
      } else if (c === '}' || c === ']') {
        depth--;
      }
    }
  }
}

/**
 * Custom reviver function for JSON.parse to remove prototype pollution keys.
 * @param key - The property key.
 * @param value - The property value.
 * @returns The cleaned value, or undefined if the key is prototype-polluting.
 */
export function safeJsonReviver(key: string, value: any): any {
  if (key === '__proto__' || key === 'constructor' || key === 'prototype') {
    return undefined;
  }
  return value;
}

/**
 * Safely parses a JSON string, enforcing nesting limits and stripping prototype pollution.
 * @param text - The JSON string to parse.
 * @param options - Configuration options such as maxDepth.
 * @returns The parsed and sanitized value.
 * @throws VibeShieldDeserializationError on nesting limit exceeded or malformed input.
 */
export function safeJsonParse<T>(text: string, options?: { maxDepth?: number }): T {
  const maxDepth = options?.maxDepth ?? 64;
  checkJsonDepth(text, maxDepth);

  try {
    const parsed = JSON.parse(text, safeJsonReviver);
    
    // Extra guard: recursively clean up objects just in case JSON.parse reviver missed anything due to platform differences
    const cleanObject = (obj: any): any => {
      if (obj === null || typeof obj !== 'object') return obj;
      if (Array.isArray(obj)) {
        return obj.map(cleanObject);
      }
      
      const clean: any = {};
      for (const k of Object.keys(obj)) {
        if (k === '__proto__' || k === 'constructor' || k === 'prototype') {
          continue;
        }
        clean[k] = cleanObject(obj[k]);
      }
      return clean;
    };

    return cleanObject(parsed);
  } catch (e: any) {
    if (e instanceof VibeShieldDeserializationError) throw e;
    throw new VibeShieldDeserializationError(`Malformed JSON payload: ${e.message}`);
  }
}

/**
 * Statically scans code for unsafe deserialization patterns.
 * @param code - The JavaScript/TypeScript source code to scan.
 * @returns The detection results including vulnerability findings.
 */
export function detectUnsafeDeserialization(code: string): DetectionResult {
  const findings: Finding[] = [];
  if (!code) return { safe: true, findings };

  const lines = code.split('\n');
  
  for (let i = 0; i < lines.length; i++) {
    const lineNum = i + 1;
    const line = lines[i];

    // Rule 1: node-serialize / serialize-to-js / funcserialize
    const libRegex = /\b(node-serialize|serialize-to-js|funcserialize)\b/;
    const libMatch = libRegex.exec(line);
    if (libMatch) {
      findings.push({
        line: lineNum,
        pattern: libMatch[1],
        severity: 'critical',
        suggestion: `Avoid using insecure deserialization library "${libMatch[1]}". Refactor to use safeJsonParse().`,
        originalCode: line.trim()
      });
    }

    // Rule 2: eval() with serialization/unserialize references
    const evalUnsafe = /\beval\s*\(\s*.*unserialize/;
    const evalMatch = evalUnsafe.exec(line);
    if (evalMatch) {
      findings.push({
        line: lineNum,
        pattern: 'eval() with unserialize',
        severity: 'critical',
        suggestion: 'Do not use eval() to execute serialized payloads.',
        originalCode: line.trim()
      });
    }
  }

  // Sort by line number
  findings.sort((a, b) => a.line - b.line);

  return {
    safe: findings.length === 0,
    findings
  };
}
